packages/smart-vaults/src/library/MultiSignerSignatureLib.sol

```diff
@@ -135,37 +134,45 @@ library MultiSignerSignatureLib {
         }
     }
 
-    function getSignerAtIndex(bytes memory signers_, uint8 index_) internal pure returns (bytes memory, uint8) {
-        uint256 start = uint256(index_) * SIGNER_SIZE;
-
-        uint8 signerType;
-        assembly {
-            signerType := byte(0, mload(add(signers_, add(32, start))))
-        }
-
-        uint256 returnLength;
-        if (signerType == EMPTY_SIGNER_TYPE || signerType == REMOVED_SIGNER_TYPE) {
-            return (new bytes(0), signerType);
-        } else if (signerType == EOA_SIGNER_TYPE) {
-            returnLength = MultiSignerLib.EOA_SIGNER_SIZE;
-        } else if (signerType == PASSKEY_SIGNER_TYPE) {
-            returnLength = MultiSignerLib.PASSKEY_SIGNER_SIZE;
-        }
-
-        bytes memory signer = new bytes(returnLength);
+    function getSignerAtIndex(bytes memory signerUpdates_, uint8 index_) internal pure returns (bytes memory) {
+        uint256 numUpdates = signerUpdates_.length;
 
-        assembly {
-            let dataPtr := add(add(signers_, 33), start)
-            let itemPtr := add(signer, 32)
+        uint8 currentIndex;
+        for (uint256 i = 0; i < numUpdates; i += SIGNER_SIZE) {
+            assembly {
+                currentIndex := byte(0, mload(add(signerUpdates_, add(32, i))))
+            }
 
-            switch returnLength
-            case 32 { mstore(itemPtr, mload(dataPtr)) }
-            case 64 {
-                mstore(itemPtr, mload(dataPtr))
-                mstore(add(itemPtr, 32), mload(add(dataPtr, 32)))
+            if (currentIndex == index_) {
+                uint8 signerType;
+                uint256 returnLength;
+
+                assembly {
+                    signerType := byte(0, mload(add(add(signerUpdates_, add(32, i)), 1)))
+                }
+
+                if (signerType == EOA_SIGNER_TYPE) {
+                    returnLength = MultiSignerLib.EOA_SIGNER_SIZE;
+                } else if (signerType == PASSKEY_SIGNER_TYPE) {
+                    returnLength = MultiSignerLib.PASSKEY_SIGNER_SIZE;
+                }
+
+                bytes memory signer = new bytes(returnLength);
+                assembly {
+                    let dataPtr := add(add(signerUpdates_, i), 34) // Skip index and signer type bytes
+                    let destPtr := add(signer, 32)
+
+                    switch returnLength
+                    case 32 { mstore(destPtr, mload(dataPtr)) }
+                    case 64 {
+                        mstore(destPtr, mload(dataPtr))
+                        mstore(add(destPtr, 32), mload(add(dataPtr, 32)))
+                    }
+                }
+                return signer;
             }
         }
 
-        return (signer, signerType);
+        revert SignerNotPresent(index_);
```

Comment:
i think this will cause issues for dummy signatures / gas estimation?

Reply:
i think we probably need to either return a bool for success v failure or perhaps just an empty `bytes`? (w a check in `isValidSignature`)

Reply:
how would this break gas estimation/dummy signatures?